File: Day07/Hard/Solution.java
import java.util.*;

class Solution {
    public String[] spellchecker(String[] wordlist, String[] queries) {
        Set<String> exact = new HashSet<>(Arrays.asList(wordlist));
        Map<String, String> caseInsensitive = new HashMap<>();
        Map<String, String> vowelInsensitive = new HashMap<>();

        for (String word : wordlist) {
            String lower = word.toLowerCase();
            String devowel = devowel(lower);
            caseInsensitive.putIfAbsent(lower, word);
            vowelInsensitive.putIfAbsent(devowel, word);
        }

        String[] ans = new String[queries.length];
        for (int k = 0; k < queries.length; k++) {
            String q = queries[k];
            if (exact.contains(q)) {
                ans[k] = q;
                continue;
            }

            String lowerQ = q.toLowerCase();
            if (caseInsensitive.containsKey(lowerQ)) {
                ans[k] = caseInsensitive.get(lowerQ);
                continue;
            }

            String devowelQ = devowel(lowerQ);
            if (vowelInsensitive.containsKey(devowelQ)) {
                ans[k] = vowelInsensitive.get(devowelQ);
                continue;
            }

            ans[k] = "";
        }
        return ans;
    }

    private String devowel(String s) {
        return s.replaceAll("[aeiou]", "*");
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        String[] wordlist = {"KiTe","kite","hare","Hare"};
        String[] queries = {"kite","Kite","KiTe","Hare","HARE","Hear","hear","keti","keet","keto"};
        String[] ans = sol.spellchecker(wordlist, queries);

        for (String s : ans) System.out.print("\"" + s + "\" ");
        System.out.println();
    }
}

File: Day07/Hard/Solution.py
from typing import List

class Solution:
    def spellchecker(self, wordlist: List[str], queries: List[str]) -> List[str]:
        exact = set(wordlist)
        caseInsensitive = {}
        vowelInsensitive = {}

        def devowel(word: str) -> str:
            return ''.join('*' if c in "aeiou" else c for c in word.lower())

        for word in wordlist:
            lower = word.lower()
            dv = devowel(word)
            if lower not in caseInsensitive:
                caseInsensitive[lower] = word
            if dv not in vowelInsensitive:
                vowelInsensitive[dv] = word

        ans = []
        for q in queries:
            if q in exact:
                ans.append(q)
            elif q.lower() in caseInsensitive:
                ans.append(caseInsensitive[q.lower()])
            elif devowel(q) in vowelInsensitive:
                ans.append(vowelInsensitive[devowel(q)])
            else:
                ans.append("")
        return ans

if __name__ == "__main__":
    sol = Solution()
    wordlist = ["KiTe","kite","hare","Hare"]
    queries = ["kite","Kite","KiTe","Hare","HARE","Hear","hear","keti","keet","keto"]
    print(sol.spellchecker(wordlist, queries))
